packages/dds/tree/src/simple-tree/treeApi.ts

```diff
@@ -141,11 +156,58 @@ export const nodeApi: TreeApi = {
 
 /**
  * A collection of events that can be raised by a {@link TreeNode}.
+ *
```

Comment:
I think I have a solution:

For our planned events (before and after change events), we currently guarantee (implicitly: we need to document this) that we will break down the edit into a sequence of changes, each wrapped with a before and after such that only the edit in question happens between its before and after, and no edits happen without being wrapped with before and after events.

To tolerate non-local atomic events like swaps, we can slightly relax this guarantee. For after events, we guarantee that the only change to the subtree under the impacted node (instead of the entire document) will the edit described by the event. We can still guarantee the whole document will be valid and ins schema, it just might have some extra edits already applied that haven't had events yet.

This means that if a swap occurs, we can do the whole edit then apply the after events (in a deeper in the tree to shallower order if one is under the other).

This approach seems like it should work pretty well for invalidation style use.

How this could be applied to before events is still an open question. Perhaps we fire both in a top-down order.

Consumers of delta style APIs might be forced to actually understand swaps, or have some sufficiently general (ex: reencode can express all tree changes including swaps), since neither the before or after events can separately really describe a swap. They could provide a non-identity preserving approximation, or an incremental change that temporally multi parents (ex: express swap as two replace operations where the source might already have a parent)